### ISSUE TITLE
Allow specification of dimensions with shape greater than one at the graph and/or `Type` levels

### ISSUE THREAD

**Issue by brandonwillard** [aesara-devs/aesara#1122]
### Discussed in https://github.com/aesara-devs/aesara/discussions/1104

<div type='discussions-op-text'>

<sup>Originally posted by **ricardoV94** August  8, 2022</sup>
Brought up in #1089 

Knowing that two tensors will not broadcast (i.e., dimensions > 1) allows for more succinct graphs in some cases as well assumptions in rewrite. In fact many old rewrites developed in Theano relied on something like that (although it was a bit murky, see #1089 for more details), and are nowadays incorrect. Here is one known example in the codebase:

 https://github.com/aesara-devs/aesara/blob/7f8af9bc28755d93dca3afff2534a8a5f5ecbd80/aesara/tensor/basic_opt.py#L1695-L1705

A common pattern, is that when you have an Elemwise operation between two inputs, and they have the same broadcastable pattern, you can assume they have the same shape (otherwise it would be an invalid graph). With current Aesara Types you can not make this assumption because there is no way to specify `shape>1` at the Type level, without specifying a specific shape. </div>

**Comment by brandonwillard** [comment]
To move forward on this sooner than later, we could always consider allowing `-1` to encode a "not equal to 1" (i.e. not broadcastable) constraint on `TensorType.shape` values.  I believe this has come up before, and it isn't perhaps the most desirable approach, but it should work to adequately express the old `TensorType.broadcastable` constraints while also allowing `TensorType.shape[d] == None` to represent a complete lack of shape information&mdash;as it intuitively does.

We would need to refactor our current usage of `TensorType.shape` to account for this case, but, aside from being a less than exciting task, it shouldn't be particularly difficult.

Making this change would finally allow us to reason more definitively about compile-time broadcasting, and all without the strong and unreliable assumptions we are currently required to make in our constructors and rewrites.  More specifically, it would allow us to raise an error in the example of https://github.com/aesara-devs/aesara/issues/1089 when the gradient is constructed.

**Comment by rlouf** [comment]
Discussions in #1089 are hard to follow, and I have a naive question: why do we need to add a broadcastability constraints if we can do shape inference? Wouldn't it be enough to have something like `np.broadcast_shapes` as part of shape inference, in addition to a convention to represent unknown ranks/shapes?

**Comment by brandonwillard** [comment]
> Discussions in #1089 are hard to follow, and I have a naive question: why do we need to add a broadcastability constraints if we can do shape inference? Wouldn't it be enough to have something like `np.broadcast_shapes` as part of shape inference, in addition to a convention to represent unknown ranks/shapes?

Part of the issue is that shape inference (i.e. `ShapeFeature` and the associated rewrites) only really exists at compile-time (i.e. within our `FunctionGraph` and `Function` framework), yet issues like #1089 involve graphs that are constructed in places like `Op.make_node`, `Op.grad`, etc., which are used at "construction-time" (e.g. when a user constructs a graph).  In those cases, we have no access to the shape inference framework.  The only (shape/broadcast) information we have to work with is at the `Type`-level (i.e. in the `TensorType` objects) and that information is almost exclusively provided by and maintained by the `Op.make_node` implementations.

If we add a constraint to represent partial shape information, then we can continue to propagate and distinguish between the same information provided by `TensorType.broadcastable` and the "no shape information/`None`" case in `TensorType.shape`.  Shape inference can also use that same information to its advantage, though, so it's a way to move forward in more than one direction, as well.  I would much rather keep all of this in one place (i.e. compile-time), but the design of `aesara.grad` and its use of `Op.grad` doesn't allow that at the moment.  As has been mentioned before, if we turned gradients into purely symbolic `Op`s and then expanded them at compile-time, then problems in #1089 would be entirely within compile-time scope.

**Comment by brandonwillard** [comment]
> The only (shape/broadcast) information we have to work with is at the `Type`-level (i.e. in the `TensorType` objects) and that information is almost exclusively provided by and maintained by the `Op.make_node` implementations.

By the way, this is also one of the primary reasons for actually supporting the "dynamic" broadcasting (i.e. broadcasting without the "required" static shape information) mentioned in #1089: we can't rely on every `Op.make_node` to correctly propagate that static shape information.  Sometimes that's due to a particularly lacking implementation, sometimes it's due to a lack of requisite information in that context (e.g. more elaborate shape inference is needed), and other times it could be due to a definitional inability to make such shape inferences for a specific `Op`.

If we simply say "no, we don't support this" across the board, then broadcasting will break in graphs containing such `Op`s, and I think we would rather have those graphs work whenever possible.  Perhaps we'll need to say, "no, we don't support _gradients_" in these cases, but we definitely _shouldn't_ say we don't support it when we absolutely *can*.
